common/djangoapps/student/roles.py

```diff
@@ -60,21 +62,47 @@ def strict_role_checking():
     ACCESS_ROLES_INHERITANCE.update(OLD_ACCESS_ROLES_INHERITANCE)
 
 
+def get_role_cache_course_key(course_id=None):
+    """
+    Get the cache key for the course key.
+    """
+    default = ROLE_CACHE_UNGROUPED_COURSES_KEY
+    if not course_id:
+        return default
+    if not getattr(course_id, 'html_id', False):
+        return default
+    html_id = course_id.html_id()
+    if not html_id:
+        return default
+
+    return html_id
+
+
 class BulkRoleCache:  # lint-amnesty, pylint: disable=missing-class-docstring
     CACHE_NAMESPACE = "student.roles.BulkRoleCache"
     CACHE_KEY = 'roles_by_user'
 
     @classmethod
     def prefetch(cls, users):  # lint-amnesty, pylint: disable=missing-function-docstring
-        roles_by_user = defaultdict(set)
+        roles_by_user = {}
         get_cache(cls.CACHE_NAMESPACE)[cls.CACHE_KEY] = roles_by_user
 
         for role in CourseAccessRole.objects.filter(user__in=users).select_related('user'):
-            roles_by_user[role.user.id].add(role)
+            user_id = role.user.id
+            course_id = get_role_cache_course_key(role.course_id)
+            if not roles_by_user.get(user_id):
+                roles_by_user[user_id] = {course_id: set()}
+            if not roles_by_user[user_id].get(course_id):
+                roles_by_user[user_id][course_id] = set()
+            roles_by_user[user_id][course_id].add(role)
+
+            if not roles_by_user.get(course_id):
```

Comment:
I think I made quite a mistake here. I think it should be `if not roles_by_user[user_id].get(course_id)`. What's even the purpose of this since the three lines above already cover it? I think I did something wrong here. I'll have another look.

Reply:
done (removed these lines)